File: LeetCode/Python Solutions/Find Largest Value in Each Tree Row/find_largest_value_in_each_tree_row.py
from collections import deque

class Solution:
    def largestValues(self, root):
        if not root:
            return []
        
        # Initialize the result list and the queue for level-order traversal
        result = []
        queue = deque([root])
        
        while queue:
            # Start the current level
            level_size = len(queue)
            max_value = float('-inf')  # Initialize the maximum value for this level
            
            # Process all nodes in the current level
            for _ in range(level_size):
                node = queue.popleft()
                max_value = max(max_value, node.val)  # Update the maximum value
                
                # Add children to the queue for the next level
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
            
            # After finishing the level, add the maximum value to the result list
            result.append(max_value)
        
        return result

File: LeetCode/Java Solutions/Find Largest Value in Each Tree Row/find_largest_value_in_each_tree_row.java
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

public class Solution {
    public List<Integer> largestValues(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        if (root == null) {
            return result;
        }

        // Initialize the queue for level-order traversal
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);

        while (!queue.isEmpty()) {
            int levelSize = queue.size();
            int max = Integer.MIN_VALUE;  // Initialize the maximum value for this level

            // Process all nodes in the current level
            for (int i = 0; i < levelSize; i++) {
                TreeNode node = queue.poll();
                max = Math.max(max, node.val);  // Update the maximum value

                // Add children to the queue for the next level
                if (node.left != null) {
                    queue.offer(node.left);
                }
                if (node.right != null) {
                    queue.offer(node.right);
                }
            }

            // After finishing the level, add the maximum value to the result list
            result.add(max);
        }

        return result;
    }
}
